File: internal/content/story.go
package content

import (
	"fmt"
	"io/fs"
	"path"
	"sort"

	"gopkg.in/yaml.v3"

	"github.com/sebastiancaraballo/polyglot/internal/model"
)

// storyFile mirrors the on-disk YAML shape of a story chapter.
type storyFile struct {
	ID    string     `yaml:"id"`
	Title string     `yaml:"title"`
	Beats []beatFile `yaml:"beats"`
}

// beatFile mirrors the on-disk YAML shape of a single beat.
type beatFile struct {
	Kind     string `yaml:"kind"`
	Speaker  string `yaml:"speaker"`
	Place    string `yaml:"place"`
	Source   string `yaml:"es"`
	JP       string `yaml:"jp"`
	Romaji   string `yaml:"romaji"`
	Practice string `yaml:"practice"`
	RefID    string `yaml:"ref_id"`
}

// loadChapters reads every story chapter for pair. Like grammar patterns,
// story content is optional: a pair with no content/<pair>/story directory
// yet is not an error, since Core must not force every language pair to ship
// story content on day one.
func loadChapters(fsys fs.FS, pair string) ([]model.Chapter, error) {
	dir := path.Join("content", pair, "story")
	files, err := fs.Glob(fsys, path.Join(dir, "*.yaml"))
	if err != nil {
		return nil, fmt.Errorf("glob story: %w", err)
	}
	sort.Strings(files)

	seen := make(map[string]bool)
	var chapters []model.Chapter
	for _, file := range files {
		c, err := parseChapter(fsys, file)
		if err != nil {
			return nil, err
		}
		if seen[c.ID] {
			return nil, fmt.Errorf("%s: duplicate chapter id %q", file, c.ID)
		}
		seen[c.ID] = true
		chapters = append(chapters, c)
	}
	return chapters, nil
}

func parseChapter(fsys fs.FS, file string) (model.Chapter, error) {
	data, err := fs.ReadFile(fsys, file)
	if err != nil {
		return model.Chapter{}, fmt.Errorf("read %s: %w", file, err)
	}
	var sf storyFile
	if err := yaml.Unmarshal(data, &sf); err != nil {
		return model.Chapter{}, fmt.Errorf("parse %s: %w", file, err)
	}
	if sf.ID == "" {
		return model.Chapter{}, fmt.Errorf("%s: missing chapter id", file)
	}
	if sf.Title == "" {
		return model.Chapter{}, fmt.Errorf("%s: missing chapter title", file)
	}
	if len(sf.Beats) == 0 {
		return model.Chapter{}, fmt.Errorf("%s: chapter has no beats", file)
	}

	beats := make([]model.Beat, 0, len(sf.Beats))
	for i, b := range sf.Beats {
		beat, err := parseBeat(b)
		if err != nil {
			return model.Chapter{}, fmt.Errorf("%s: beat %d: %w", file, i+1, err)
		}
		beats = append(beats, beat)
	}

	return model.Chapter{ID: sf.ID, Title: sf.Title, Beats: beats}, nil
}

func parseBeat(b beatFile) (model.Beat, error) {
	kind := model.BeatKind(b.Kind)
	if !kind.Valid() {
		return model.Beat{}, fmt.Errorf("invalid beat kind %q", b.Kind)
	}

	beat := model.Beat{Kind: kind, Speaker: b.Speaker, Place: b.Place}

	switch kind {
	case model.Narration, model.Dialogue:
		if b.Source == "" || b.JP == "" {
			return model.Beat{}, fmt.Errorf("%s beat is missing es or jp", kind)
		}
		if kind == model.Dialogue && b.Speaker == "" {
			return model.Beat{}, fmt.Errorf("dialogue beat is missing speaker")
		}
		beat.Source, beat.JP, beat.Romaji = b.Source, b.JP, b.Romaji
	case model.Present:
		practice := model.PracticeKind(b.Practice)
		if !practice.Valid() {
			return model.Beat{}, fmt.Errorf("invalid practice kind %q", b.Practice)
		}
		if b.RefID == "" {
			return model.Beat{}, fmt.Errorf("present beat is missing ref_id")
		}
		// The framing line (speaker/es/jp/romaji) is optional: a present beat
		// may set the scene diegetically, but its core content is the pool it
		// renders. es/jp travel together when given.
		if (b.Source == "") != (b.JP == "") {
			return model.Beat{}, fmt.Errorf("present beat framing needs both es and jp, or neither")
		}
		beat.Practice, beat.RefID = practice, b.RefID
		beat.Source, beat.JP, beat.Romaji = b.Source, b.JP, b.Romaji
	case model.Practice:
		practice := model.PracticeKind(b.Practice)
		if !practice.Valid() {
			return model.Beat{}, fmt.Errorf("invalid practice kind %q", b.Practice)
		}
		if b.RefID == "" {
			return model.Beat{}, fmt.Errorf("practice beat is missing ref_id")
		}
		if b.Source != "" || b.JP != "" || b.Romaji != "" || b.Speaker != "" {
			return model.Beat{}, fmt.Errorf("practice beat must not carry dialogue fields")
		}
		beat.Practice, beat.RefID = practice, b.RefID
	}
	return beat, nil
}

File: core/src/content/story.rs
use std::collections::HashSet;

use serde::Deserialize;

use super::fsys::ContentFs;
use super::LoadError;
use crate::model::{Beat, BeatKind, Chapter, PracticeKind};

/// Mirrors the on-disk YAML shape of a story chapter.
#[derive(Deserialize)]
struct StoryFile {
    #[serde(default)]
    id: String,
    #[serde(default)]
    title: String,
    #[serde(default)]
    beats: Vec<BeatFile>,
}

/// Mirrors the on-disk YAML shape of a single beat.
#[derive(Deserialize)]
struct BeatFile {
    #[serde(default)]
    kind: String,
    #[serde(default)]
    speaker: String,
    #[serde(default)]
    place: String,
    #[serde(default, rename = "es")]
    source: String,
    #[serde(default)]
    jp: String,
    #[serde(default)]
    romaji: String,
    #[serde(default)]
    practice: String,
    #[serde(default)]
    ref_id: String,
}

/// Reads every story chapter for `pair`. Like grammar patterns, story content
/// is optional.
pub(super) fn load_chapters(fsys: &dyn ContentFs, pair: &str) -> Result<Vec<Chapter>, LoadError> {
    let dir = format!("{pair}/story");
    let files = fsys
        .glob_yaml(&dir)
        .map_err(|e| LoadError::new(format!("glob story: {e}")))?;

    let mut seen: HashSet<String> = HashSet::new();
    let mut chapters = Vec::new();
    for file in files {
        let c = parse_chapter(fsys, &file)?;
        if !seen.insert(c.id.clone()) {
            return Err(LoadError::new(format!(
                "{file}: duplicate chapter id {:?}",
                c.id
            )));
        }
        chapters.push(c);
    }
    Ok(chapters)
}

fn parse_chapter(fsys: &dyn ContentFs, file: &str) -> Result<Chapter, LoadError> {
    let data = fsys
        .read(file)
        .map_err(|e| LoadError::new(format!("read {file}: {e}")))?;
    let sf: StoryFile = serde_yaml_ng::from_slice(&data)
        .map_err(|e| LoadError::new(format!("parse {file}: {e}")))?;

    if sf.id.is_empty() {
        return Err(LoadError::new(format!("{file}: missing chapter id")));
    }
    if sf.title.is_empty() {
        return Err(LoadError::new(format!("{file}: missing chapter title")));
    }
    if sf.beats.is_empty() {
        return Err(LoadError::new(format!("{file}: chapter has no beats")));
    }

    let mut beats = Vec::with_capacity(sf.beats.len());
    for (i, b) in sf.beats.iter().enumerate() {
        let beat =
            parse_beat(b).map_err(|e| LoadError::new(format!("{file}: beat {}: {e}", i + 1)))?;
        beats.push(beat);
    }

    Ok(Chapter {
        id: sf.id,
        title: sf.title,
        beats,
    })
}

fn parse_beat(b: &BeatFile) -> Result<Beat, String> {
    let Some(kind) = BeatKind::from_str(&b.kind) else {
        return Err(format!("invalid beat kind {:?}", b.kind));
    };

    let mut beat = Beat {
        kind,
        speaker: b.speaker.clone(),
        place: b.place.clone(),
        source: String::new(),
        jp: String::new(),
        romaji: String::new(),
        practice: None,
        ref_id: String::new(),
    };

    match kind {
        BeatKind::Narration | BeatKind::Dialogue => {
            if b.source.is_empty() || b.jp.is_empty() {
                return Err(format!("{} beat is missing es or jp", kind.as_str()));
            }
            if kind == BeatKind::Dialogue && b.speaker.is_empty() {
                return Err("dialogue beat is missing speaker".to_string());
            }
            beat.source = b.source.clone();
            beat.jp = b.jp.clone();
            beat.romaji = b.romaji.clone();
        }
        BeatKind::Present => {
            let Some(practice) = PracticeKind::from_str(&b.practice) else {
                return Err(format!("invalid practice kind {:?}", b.practice));
            };
            if b.ref_id.is_empty() {
                return Err("present beat is missing ref_id".to_string());
            }
            // The framing line is optional, but es/jp travel together.
            if b.source.is_empty() != b.jp.is_empty() {
                return Err("present beat framing needs both es and jp, or neither".to_string());
            }
            beat.practice = Some(practice);
            beat.ref_id = b.ref_id.clone();
            beat.source = b.source.clone();
            beat.jp = b.jp.clone();
            beat.romaji = b.romaji.clone();
        }
        BeatKind::Practice => {
            let Some(practice) = PracticeKind::from_str(&b.practice) else {
                return Err(format!("invalid practice kind {:?}", b.practice));
            };
            if b.ref_id.is_empty() {
                return Err("practice beat is missing ref_id".to_string());
            }
            if !b.source.is_empty()
                || !b.jp.is_empty()
                || !b.romaji.is_empty()
                || !b.speaker.is_empty()
            {
                return Err("practice beat must not carry dialogue fields".to_string());
            }
            beat.practice = Some(practice);
            beat.ref_id = b.ref_id.clone();
        }
    }
    Ok(beat)
}
